feat(previews): previews for large remote files without full file download

invario opened the pull request:
In a nutshell, previews for remote videos were partially disabled for any files larger than 5MB for performance reasons.  (To avoid the situation of say...downloading a 37GB video file from S3.)

(I have a separate pending PR #53634 for S3 that allows preview generation via a direct connection to S3)

See the following for more discussion and detail of the problem this addresses:
* Resolves: # <!-- related github issue -->
#53469 
#53496 
#52079 

## Summary
Thanks to an idea from Derkedes in this post: https://github.com/nextcloud/server/issues/53469#issuecomment-3065047547

> Downloading only a small part of the video is a useful optimization, however it also needs to support video formats that place important data at the end of the file. Instead of downloading only the first 5MB, would be possible to create a sparse file and download the first and last 5MB? The middle of the file would be empty, but ffmpeg should be able to extract a single frame.
> 
> I think this issue should be classified as a bug, since video thumbnails are now broken for mp4 files on external storages.

**This PR does the following:**

- It will search for the `moov atom` in the original MP4/MOV file
- It will create a blank file of same size as the video
- It will copy the first 10MB of the remote file to the beginning of the new sparse file
- It will copy the `moov atom` that it located to the same offset in the new sparse file
- This sparse file will be used to generate a preview using an image grab from 5 sec and (failing that) also 1, and 0 seconds in.
- Adds a new test `MovieTestRemoteFile` to verify that sparse file generation is working by comparing to a reference preview 
- Adds a new higher bitrate video to use for testing

**Important notes:**

- I increased the download amount from 5MB to 10MB for the leading download. This is necessary because 5MB is too little to achieve reliable preview generation from modern video files.
- For reference, 1080p H265/HEVC 30FPS recorded on a GS25+ has a bitrate of about 10Mbps.  This translates to requiring 6.25 MB for 5 seconds of video. 4K H265/HEVC 60FPS is 70Mbps, which translates to 44 MB for 5 seconds of video.
- Under this PR's sparse file method, `ffmpeg` has returned an error when attempting to generate a preview/still using a timestamp that requires data from the empty/null part of the sparse file. This is good and expected behavior and will result in fallback from 5 to 1 to 0 seconds.

## TODO

- [ ] Testers?  I have tested this on an external SMB storage mount using files as large as 2GB.
- [ ] Input and discussion?

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +99,33 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		if (($size+5242880) >= $file->getSize()) {
+			stream_copy_to_stream($content, $sparseFile);
+		} else {
+			stream_copy_to_stream($content, $sparseFile, $size);
+			stream_copy_to_stream($zeroes, $sparseFile, ($file->getSize()-($size+5242880)));
+			stream_copy_to_stream($content, $sparseFile, 5242880, ($file->getSize()-5242880));
```
> But it occurs to me that this might be dependent on the capabilities of the external storage wrapper?

Yes, that is the problem. We'd need to make sure that all of them support it or if one of them doesn't the approach for it specifically needs to be changed (probably to what we had before).

invario replied:
Modified the code to check `stream_get_meta_data` for the `[seekable]` value and it seems to be returning `true` so I have made the changes in the code to verify against that first and fallback to normal/existing routines.

invario commented:
> Wow, thank you for turning my suggestion into actual code, I did not expect this! I want to clarify one thing about sparse files. A [sparse file](https://en.wikipedia.org/wiki/Sparse_file) is a file with a "hole"; you do not need to fill the middle with zeroes. If you write zeroes to the middle of the file, it is not a sparse file anymore.
> 
> I am not very familiar with PHP as I am with other languages, but here's an untested example:
> 
> ```
> $input = ...
> $output = ...
> $total_size = 10*1024*1024*1024; // 10GiB
> $partial_size = 5242880; // 5MiB
> 
> ftruncate($output, $total_size); // create a large file without actually writing any data to disk
> // Copy first part to start of file
> fseek($output, 0);
> stream_copy_to_stream($input, $output, $partial_size, 0);
> // Copy last part to end of file
> fseek($output, $total_size - $partial_size);
> stream_copy_to_stream($input, $output, $partial_size, $total_size - $partial_size);
> ```

Thanks! So I am (incorrectly) saying it is zerose.  [/dev/zero](https://en.wikipedia.org/wiki//dev/zero) actually provides null zeroes.  I quickly generated a file using ftruncate to the same size as another file I generated using /dev/zero with 0 diffs between them.

invario commented:
> > I quickly generated a file using ftruncate to the same size as another file I generated using /dev/zero with 0 diffs between them.
> 
> Indeed, there is no difference between the file content. But if you copy from /dev/zero to the file, actual data is written to disk.
> 
> Example:
> 
> ```
> dd if=/dev/zero of=testfile1 bs=1M count=100
> truncate -s 100M testfile2
> ```
> 
> Both files appear to be 100MB of zero data:
> 
> ```
> ls -l testfile*
> -rw-r--r--. 1 robin robin 104857600 Jul 16 16:17 testfile1
> -rw-r--r--. 1 robin robin 104857600 Jul 16 16:17 testfile2
> ```
> 
> But one file consumes 100MB of disk space, and the other 0 bytes (only metadata):
> 
> ```
> du testfile*
> 102404  testfile1
> 0       testfile2
> ```
> 
> It will be a lot faster to use ftruncate and fseek, instead of writing from /dev/zero to file file. Using ftruncate, you only need to write 2x 5MB of data. If you copy /dev/zero to the file, you many need to write many gigabytes of zeros to the file.

Understood, thanks for the explanation! Will implement this.

invario commented:
Spoke too soon, it was my mistake.  `[seekable]` seems to be returning `true` so I have made the changes in the code to check first.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -78,13 +89,8 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
-					$result = $this->generateThumbNail($maxX, $maxY, $absPath, 0);
-				}
-			}
+			// Attempt still image grab from 1 second and 0 second timestamp
+			$result = $this->generateThumbNail($maxX, $maxY, $absPath, 1) ?? $this->generateThumbNail($maxX, $maxY, $absPath, 0);
```
Why not keep the attempt for 5 seconds? It probably generates better previews and will likely work on many videos.

provokateurin replied:
Ah so it could happen that we get a broken preview? I thought in this case null would be returned, so that 1 and 0 are attempted as well.
I wonder if there would be any way to figure out the bitrate of a video and then download just the right amount of data to generate a preview from the 5 second timestamp? This might get out of scope for this PR though, I just wanted to leave that as an idea.

provokateurin replied:
Is there any way we could just offload this to ffmpeg instead of replicating the behavior? I don't think it's a good idea to implement all of it in Nextcloud.

invario replied:
I have a separate PR #53634 that does that for S3 buckets. (Generates a pre-signed URI and passes that URI directly to ffmpeg). But that only works because the URI is presigned and because ffmpeg supports http/https. ffmpeg doesn't support SMB on most packages. `ffmpeg -protocols` lists the following in my docker container:
```
Input:
  async
  bluray
  cache
  concat
  concatf
  crypto
  data
  fd
  ffrtmpcrypt
  ffrtmphttp
  file
  ftp
  gopher
  gophers
  hls
  http
  httpproxy
  https
  mmsh
  mmst
  pipe
  rtmp
  rtmpe
  rtmps
  rtmpt
  rtmpte
  rtmpts
  rtp
  srtp
  subfile
  tcp
  tls
  udp
  udplite
  unix
  rist
  srt
  sftp
  zmq
  ipfs
  ipns

```

provokateurin replied:
It seems to be supported, but needs to be enabled at compile time and I guess in many cases it's not done by the distros: https://ffmpeg.org/ffmpeg-all.html#libsmbclient

I think it would be fine to say that ffmpeg should just handle it all and if you want previews for smb storages then you need ffmpeg with smb support (this should probably be a setup check to ensure that for each external storage the used ffmpeg has the appropriate support compiled).

The harder part is probably to always pass all the right options to ffmpeg, so that it will work in all cases (e.g. custom certificates, ports and so on).

invario replied:
I'm not sure if even if SMB were compiled in it would be possible for ffmpeg to do so directly as the URI would be missing authentication and it strikes me that passing username and password directly to ffmpeg like so would be a bad idea: `smb://USER:PASSWORD@myserver://share/folder`

invario replied:
ffmpeg docs say yes, but quick googling says that input formats have to be explicitly specified to work. Nevertheless, I tried it and got this from the command line:
`[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0x7f080a76d940] Error during demuxing: Invalid data found when processing input`

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +101,42 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+
+		$sparseFile = fopen($absPath, 'w');
+		// If filesize is small (i.e. <= $size + 5 MB) then just download entire file
+		if (($size + 5242880) >= $file->getSize()) {
```
Could you create a constant for 5242880 (and 10485760) so that they are not copied over and over? Also I would prefer if that constant is calculated at runtime, so it's easily verifiable that the value is correct.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +101,42 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+			// Create a sparse file of equal size to original video
+			ftruncate($sparseFile, $file->getSize());
+			// Copy $size bytes to front end
+			fseek($sparseFile, 0);
```
Is this first fseek actually necessary?

Derkades replied:
Yes. 

> You MUST use rewind() after ftruncate() to replace file content

From a comment on https://www.php.net/manual/en/function.ftruncate.php

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +101,42 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
 		return $result;
 	}
+	
+	private function getSparseFile(File $file, int $size): string|false {
```
So there is $size which is the number of bytes from the beginning of the file, but you also have 5MB from the end of the file. I think it would make more sense if both just use $size.

invario replied:
I decoupled the values because the 5MB from the end is to obtain the `moov atom` whereas the data from the front is actual video (and the `moov atom if present at front).  The `moov atom` can be as large as 32 MB for a 9 hour long video (allegedly, came across this while searching Google) but for reference: 
In my 4K HEVC video of 9 secs, the `moov atom` is only 12 KB.  On a 1080p x264 video of 1.5 hours, the `moov atom` is 3 MB.

invario commented:
@provokateurin Force pushed an update. Couldn't move the `mov atom` to the front like I originally planned (it's not as simple as moving it, and ffprobe/ffmpeg refuse to process it.  So I stuck with generating a sparse file. Added lots of comments to explain what's going on.

Also re-added in previews for the 5 sec timestamp for local files and limited the 1 and 0 sec ones for ones that require temp files.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +116,101 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		$sparseFile = fopen($absPath, 'w');
+		
+		// If video size is less than or equal to $size then just download entire file
+		if (($size) >= $file->getSize()) {
```
```suggestion
		if ($size >= $file->getSize()) {
```

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +116,101 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+					// Found the 'moov' atom, store its location and size
+					if ($atomLabel === 'moov') {
+						$moovSize = $atomSize;
+						$moovOffset = $offset;
```
Maybe do an explicit break here. The while loop should already end if this case is triggered, but with a break it should be a bit clearer.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +116,101 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+				}
+			} else {
+				// 'ftyp' atom not found, not a valid MP4/MOV
+				fclose($content);
```
```suggestion
				fclose($content);
				fclose($sparseFile);
```

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -95,6 +116,101 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+				// Capping it at $size is a precaution against a corrupt/malicious 'moov' atom
+				// Also, if the 'moov' atom size+offset extends past EOF, it is invalid.
+				if (($moovSize === 0) || ($moovSize > $size) || ($moovOffset + $moovSize > $file->getSize())) {
+					return false;
```
```suggestion
					fclose($content);
					fclose($sparseFile);
					return false;
```

provokateurin reviewed (changes requested):
Thanks for all the code comments, this was really useful to understand everything!

Would you be able to add some tests for this including real videos for testing the different scenarios? That would be really nice to ensure we don't have any regressions in the future.

Also once this PR is merged it would be cool to have an issue about the idea to let ffmpeg handle all of it. I still think we could make it work somehow and make it more reliable and effective that way.

invario commented:
> Thanks for all the code comments, this was really useful to understand everything!
> 
👍 

> Would you be able to add some tests for this including real videos for testing the different scenarios? That would be really nice to ensure we don't have any regressions in the future.
> 

I don't know how to do that but I'll work on figuring it out.  On a related note, I came across this video in the `tests` directory from 10 years ago.
https://github.com/nextcloud/server/tree/master/tests/data
It's no wonder 5MB was the amount used for the initial grab and 5 seconds for an initial attempt; this file is only 375 KB total for the entire 5 seconds, resolution of 560x320 and a bitrate of only 465 KB/s.  Times sure have changed.

> Also once this PR is merged it would be cool to have an issue about the idea to let ffmpeg handle all of it. I still think we could make it work somehow and make it more reliable and effective that way.
👍

invario commented:
- Updated `MovieTest.php` to add a new test using a mock remote file
- Updated `testimage.mp4` to be a large and high bitrate video file that is a more accurate representation of what users may be storing nowadays. I used a 4K 60FPS HDR HEVC/H265 10 second video taken from my Galaxy S25+ phone. I'm open to suggestions for if this file is too large, too long, etc.
- Fixed getSparseFile method (I had it checking a stream's `seekable` flag to be not `false` instead of explicitly checking for `true`
- Shuffled a few lines of code to `return false` before even attempting to get a temp file

Just to be clear, preview generation can still break without error if the timestamp chosen for remote generation exceeds what was grabbed (10 MB currently.)  I can't think of a way to protect against that, or even display a warning.

invario commented on tests/lib/Preview/MovieTest.php:
```diff
@@ -20,24 +22,75 @@
+		$preview = $provider->getThumbnail($image, $this->maxWidth, $this->maxHeight, $this->scalingUp);
+
+		$this->assertNotEquals(false, $preview);
+		$this->assertEquals(true, $preview->valid());
```
I'll test it against a few random versions of ffmpeg to see what results.  That was one of my concerns, but the other is that the `$timestamp` might be increased in the future but with a corresponding increase to `$size`, and that would cause a test failure, so would require updating.

But it's better than nothing though, and it **should** be trivial to update the test still image.

invario replied:
Ideally we need to generate a reference test still using video `$timeAttempts=[1,0]` but have the entire file be fed to ffmpeg (treat it as local).  Then compare it to the one being treated as remote. I

Not sure how to implement it because we have `$timeAttempts=[5,1,0]` for local files.

What do you think about removing the 5?

provokateurin replied:
> What do you think about removing the 5?

You shouldn't bend your implementation to satisfy the tests, so I'd say you just compare the remote to the local one for 0 and 1 and ignore 5 entirely.

invario replied:
So here's what I did and I have no idea if it's a good idea or bad, but I reverted `MovieTest.php` and instead created my own `MovieTestRemoteHBR.php` which will utilize a high bitrate video file (54.4 MB, 4K, non-HDR video, 10 secs, 41 Mbps).

`Movie.php` checks MD5 hash against the known value and forces timestamp [1, 0] to mimic produce a reference image to compare against for previews generated from the sparse file. The test will compare the two previews and if not identical, fail.

invario replied:
I had it implemented when I made this comment, but then removed it when I noticed that I couldn't force a test failure when I increased `$timestamp` to extend into the null data. That's when I realized that ffmpeg was actually returning an error if it went into the null data, which is the desired behavior.

I can still put in an image comparison if you think it's best, just to compare previews generated from local' vs 'remote'.  I don't believe I have to deal with the `timestamp` problem anymore so I'll just straight up compare MD5 hashes for the output.

invario replied:
Okay, I reimplemented the test. 2 previews are generated, one local, one remote, from timestamp=0, and their MD5 hashes are compared.

Where previews could potentially break:
If ffmpeg does not error out from trying to generate a still using the filler null data.  So far from my own testing, it **does** seem to error out, which is the desired behavior.  This will result in attempts properly being made at 5, 1, and 0 for remote files.

invario commented:
Haha, so get this.  After I put all the code in and the checks, I just figured out that it was made _**unnecessaray**_ due to the fact that I am precisely extracting the `moov atom` from the end of the video. From what I can tell, ffmpeg will simply error out if attempting to generate a preview using a `timestamp` that extends into the null data (from truncate.)

invario commented:
@provokateurin Thanks for the direction given so far. Anything else I need to work on in this PR?

provokateurin reviewed (commented):
Woops, forgot to submit my review

provokateurin reviewed (approved):
I think this is good now, just add this last suggested change and then squash all commits into a single one (don't forget to follow conventional commits and add the DCO).

invario commented:
Yikes.  Regression found.  `$file->hash('MD5')` forces a full download of the remote file. Fixing now.

come-nc commented:
@invario Can you detail how this PR and https://github.com/nextcloud/server/pull/53634 interacts?
What will happen if both get merged, which solution will be used in priority, and in which situation will the other one kick in?
This PR seems to support only mp4/mov, I think the other one is more generic?

invario commented:
@provokateurin @come-nc Anything needed from me at the moment?

come-nc commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +114,109 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		$sparseFile = fopen($absPath, 'w');
+
+		// If video size is less than or equal to $size then just download entire file
+		if ($size >= $file->getSize()) {
+			stream_copy_to_stream($content, $sparseFile);
```
This could be tested earlier and return false, no?

invario replied:
Absolutely right, I didn't see that. Thanks!

come-nc commented on lib/private/Preview/Movie.php:
```diff
@@ -53,23 +53,45 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		// Timestamps to make attempts to generate a still
+		$timeAttempts = [5, 1, 0];
+
+		// If metadata indicates this is a test, force attempts to 0
+		// to ensure stills for test and reference images come from the same
+		// timestamp
+		if ($file->getMetaData()['istest']) {
+			$timeAttempts = [0];
+		}
```
Is there no better way to avoid test-specific code in production like this?

come-nc replied:
I dunno, it’s scary.
I would prefer the test to simulate real world usage and verify it.

Either use a smaller test file, or hardcode the expected preview in the test?

come-nc reviewed (commented):
Small remarks. I think we can merge this after they are fixed.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -53,23 +53,28 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		// 4K HDR H265 60 FPS = 75 Mbps = 9 MB per second needed for a still
+		// 1080p H265 30 FPS = 10 Mbps = 1.25 MB per second needed for a still
+		// 1080p H264 30 FPS = 16 Mbps = 2 MB per second needed for a still
+		$sizeAttempts = [10485760];
```
```suggestion
		$sizeAttempts = [1024 * 1024 * 10];
```
Makes it much easier to understand the value

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +97,109 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
 	}
 
+	private function getSparseFile(File $file, int $size): string|false {
+		if ($size >= $file->getSize()) {
+			return false;
```
This returns false because the requested size is bigger than the entire file size, so that it moves on to use `getLocalFile`? I think you should either call `getLocalFile` here or add a comment that explains this.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +97,109 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+	private function getSparseFile(File $file, int $size): string|false {
+		if ($size >= $file->getSize()) {
+			return false;
+		} else {
```
```suggestion
		}
```
To save some indentation, since the previous if-clause always returns.

provokateurin commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +97,109 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+					'Failed to get sparse file to generate thumbnail for: ' . $file->getPath(),
+					['app' => 'core']
+				);
+				return false;
```
```suggestion
				fclose($content);
				return false;
```

provokateurin commented on tests/lib/Preview/MovieTestRemoteFile.php:
```diff
@@ -0,0 +1,83 @@
+<?php
+
+/**
+ * SPDX-FileCopyrightText: 2019-2024 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
 * SPDX-FileCopyrightText: 2019-2025 Nextcloud GmbH and Nextcloud contributors
```

provokateurin reviewed (changes requested):
Just some minor things that need to be adjusted, other than that this should be fine to merge now. I'll start the CI, so we can see if something breaks.

provokateurin reviewed (changes requested):
Sorry, just found some more minor changes :see_no_evil:

provokateurin commented on tests/lib/Preview/MovieTestRemoteFile.php:
```diff
@@ -0,0 +1,78 @@
+		}
+	}
+
+	/**
+	 * Launches all the tests we have
+	 *
+	 * @requires extension imagick
+	 *
+	 * @param int $widthAdjustment
+	 * @param int $heightAdjustment
+	 */
+	#[\PHPUnit\Framework\Attributes\DataProvider('dimensionsDataProvider')]
+	public function testGetThumbnail($widthAdjustment, $heightAdjustment): void {
```
Not sure, but let's not make it more complicated and just keep it like it is.

invario commented:
```
Error: lib/private/Preview/Movie.php:143:70: InvalidScalarArgument: Argument 3 of stream_get_contents expects int, but float|int provided (see https://psalm.dev/012)
Error: lib/private/Preview/Movie.php:145:51: InvalidScalarArgument: Argument 3 of stream_get_contents expects int, but float|int provided (see https://psalm.dev/012)
Error: lib/private/Preview/Movie.php:154:72: InvalidScalarArgument: Argument 3 of stream_get_contents expects int, but float|int provided (see https://psalm.dev/012)
Error: lib/private/Preview/Movie.php:179:27: InvalidScalarArgument: Argument 2 of ftruncate expects int, but float|int provided (see https://psalm.dev/012)
Error: lib/private/Preview/Movie.php:189:21: InvalidScalarArgument: Argument 2 of fseek expects int, but float|int provided (see https://psalm.dev/012)
Error: lib/private/Preview/Movie.php:190:24: InvalidScalarArgument: Argument 2 of fseek expects int, but float|int provided (see https://psalm.dev/012)
Error: lib/private/Preview/Movie.php:191:50: InvalidScalarArgument: Argument 3 of stream_copy_to_stream expects int|null, but float|int provided (see https://psalm.dev/012)```
Casting to `(int)` and will repush.

invario commented on tests/lib/Preview/MovieTestRemoteFile.php:
```diff
@@ -0,0 +1,70 @@
+		$movieBinary = $binaryFinder->findBinaryPath('avconv');
+		if (!is_string($movieBinary)) {
+			$movieBinary = $binaryFinder->findBinaryPath('ffmpeg');
+		}
```
🙀😆

invario commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +97,110 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+				return false;
+			}
+			// Generate new file of same size
+			ftruncate($sparseFile, (int)($file->getSize()));
```
Added a check on line 106 for PHP_INT_MAX and if the file size is larger than that to just return false.
if (($size >= $file->getSize()) || ($file->getSize > PHP_INT_MAX)) {

come-nc commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +97,111 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+	private function getSparseFile(File $file, int $size): string|false {
+		// File is smaller than $size or file is larger than max int size
+		// of the host so return false so getLocalFile method is used
+		if (($size >= $file->getSize()) || ($file->getSize > PHP_INT_MAX)) {
```
```suggestion
		if (($size >= $file->getSize()) || ($file->getSize() > PHP_INT_MAX)) {
```
Typo I suppose

invario committed:
feat(previews): previews and tests for large remote videos w/o full download

Co-authored-by: Kate <26026535+provokateurin@users.noreply.github.com>
Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: invario <67800603+invario@users.noreply.github.com>

invario commented on lib/private/Preview/Movie.php:
```diff
@@ -53,23 +53,28 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		// 4K HDR H265 60 FPS = 75 Mbps = 9 MB per second needed for a still
+		// 1080p H265 30 FPS = 10 Mbps = 1.25 MB per second needed for a still
+		// 1080p H264 30 FPS = 16 Mbps = 2 MB per second needed for a still
+		$sizeAttempts = [1024 * 1024 * 10];
```
Just noticed somewhere along the way I lost the test for local file to set `$sizeAttempts = null`.  I will readd the code and test ASAP.

provokateurin commented:
But let's wait until after the branch off, so it gets into Nextcloud 33, as we are in the feature freeze phase already.

Vislike commented:
Just want to throw in some cents regarding using isLocal()
As I also mentioned here https://github.com/nextcloud/server/pull/52079#issuecomment-3249294112
If you use encryption isLocal() returns false even if it is a local storage.
So this might need to be tested when using local server encryption also.

artonge commented:
Just want to see if the local PR's CI gets reflected in this PR

artonge commented:
Yeah! :partying_face:

Vislike commented on lib/private/Preview/Movie.php:
```diff
@@ -92,10 +108,111 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+					// 1 = read an additional 8 bytes after the label to get the 64 bit
+					// size of the atom. Needed for large atoms like 'mdat' (the video data)
+					if ($atomSize === 1) {
+						$atomSize = (int)hexdec(bin2hex(stream_get_contents($content, 8, (int)($offset + 8))));
```
If this expression reads a 0 from the file, $atomSize will be 0 and the expression on line 175 will read $offset += 0 so it looks like possible infinite loop

invario replied:
If it reads a 0, it reevaluates what the true atom size is which is then used on line 175.

invario replied:
Ah, you mean in case of corrupted/improper atom data? Since the 64 bit value should never be 0 in normal circumstances.

invario replied:
Got it now, instead of a break, I'll just have it return false. Thanks!
